### DISCUSSION TITLE
How to use `@Dependency` on `Codable` types: "does not conform to protocol 'Decodable'"

### DISCUSSION THREAD

**Discussion by finestructure** [pointfreeco/swift-dependencies#283] [Q&A] [answered]
Apologies if this has been asked/addressed before but I couldn't find anything in the dicussions or issues.

We've got a type `Model` that uses a dependency, let's say `\.date.now`, in a method `foo()`. This type has to be `Codable`. Attaching the `@Dependency` like so:

```swift
struct Model: Codable {
    @Dependency(\.date.now) var now
    var value: Int

    func foo() -> String { "\(value) \(now)" }
}
```

fails to compile with

```
Type 'Model' does not conform to protocol 'Decodable'
```

Do you have any guidance how to best instrument this use-case?

I can think of a couple of mechanisms:

1. Write a custom codable implementation that skips `now`:

```swift
struct Model: Codable {
    @Dependency(\.date.now) var now
    var value: Int

    private enum CodingKeys: CodingKey {
        case value
    }

    init(from decoder: any Decoder) throws {
        let container: KeyedDecodingContainer<DepTest.Model.CodingKeys> = try decoder.container(keyedBy: DepTest.Model.CodingKeys.self)

        self.value = try container.decode(Int.self, forKey: DepTest.Model.CodingKeys.value)

    }

    func encode(to encoder: any Encoder) throws {
        var container: KeyedEncodingContainer<DepTest.Model.CodingKeys> = encoder.container(keyedBy: DepTest.Model.CodingKeys.self)

        try container.encode(self.value, forKey: DepTest.Model.CodingKeys.value)
    }

    func foo() -> String { "\(value) \(now)" }
}
```

Even in this simple case it adds a lot of boilerplate that we didn't need before. Moreover, any changes to the set of properties now also requires syncing up the `Codable` conformance by hand (or by rerunning the Codable-refactor).

In our case `Model` has 30 properties and the autogenerated Codable conformance is 200 lines of code. That's not ideal to maintain 😅

2. There are some packages that add macros for customising `Codable` conformances. One of these customisation is typically to ignore certain fields. I have not yet tested this but it might work with the `@Dependency` property.

The downside is that we'd have to add another depenendency, and for a very critical part of our app.

I've seen there's the `@ObservationIgnored` macro which deals with a similar problem regarding observation. Is a similar macro as part of swift-dependencies conceivable, to deal with `Codable`?

3. Put the `@Dependency(\.date.now) var now` on another type, say `DependencyContainer`, and inject that as a parameter on the `foo()` call, as in `foo(dependencyContainer: container)`.

That would be feasible, although it would clutter up some call-sites. I guess it could be made not to be too terrible with default arguments but it still feels quite counter to the elegance of the `@Dependency` system overall.

Is there a better solution than these three? How are people dealing with `@Dependency` in the context of `Codable` types?

**Comment by mbrandonw** [comment]
Hey @finestructure, I think first and foremost we would advise against having a `@Dependency` in what should be a "plain data type." Especially codable types. When you encode the type you of course can't encode the dependency resolution, and so when it is decode it its dependencies will go back to whatever is provided by the surrounding context.

However, if you do want to continue doing this you could add a `@retroactive` conformance of `Dependency` to `Codable` that just does the trivial thing (this how `ObservationRegistrar` is handled too). You can also try moving `@Dependency` into the `foo` method, but that will subtly change how dependencies are resolved. It will resolve at the time of invoking `foo`, whereas currently dependencies are captured at the time of creating `Model`.

**Reply by finestructure** [reply]
Thanks for the quick reply, Brandon!

That's good advice not to add `@Dependency` to a `Codable` struct and I think we could avoid it entirely with some reshuffling. I _believe_ we're in an OK situation though. It's true that type is a plain data type but we're adding a helper in an extension that essentially transforms the data. One of the inputs is the current date. It feels natural to stick this in an extension on the type, because it really is independent of all other inputs except the dependency.

Either way, your suggestion to put the `@Dependency` inside the function seemed to be exactly the solution for the problem. And not just in this case. I've since found other places where we're using `@Dependency(\.date.now)` that are tricky to instrument: static functions.

I cannot put the property on the type (as a `@Dependency(\.date.now) static var now`), because that gives me Swift 6 concurrency warnings. So here as well the solution seems to be to place the `@Dependency` inside the function.

Also just generally this is quite nice, because the `@Dependency` declaration can then be very close to the call site.

If I understand you caveat regarding declaring `@Dependency` inside functions correctly, the concern is the difference between

```swift
  let model = withDependencies {
    $0.date.now = Date(timeIntervalSinceReferenceDate: 0)
  } operation: {
    Model()
  }
  
  model.foo() // <- not substituted - would this actually trigger the Unimplemented: @Dependency(\.date) failure?
```

and

```swift
  withDependencies {
    $0.date.now = Date(timeIntervalSinceReferenceDate: 0)
  } operation: {
    Model().foo() // should be ok, shouldn't it?
  }
```

Now what I'm seeing in practise is that even in case of the second variant, we're getting

```
Unimplemented: @Dependency(\.date)
```

from inside the `operation: { ... }` closure when executing the tests for `@Dependency(\.date.now) var now` declarations that are set in functions.

Am I doing this wrong? It feels like this should work as you described.

**Reply by finestructure** [reply]
FWIW, in a standalone project, this all works:

```swift
struct Model: Codable {
    func foo() -> String {
        @Dependency(\.date.now) var now
        return "\(now)"
    }

    static func foo() -> String {
        @Dependency(\.date.now) var now
        return "\(now)"
    }
}
```

```swift
import Testing
@testable import dep_test
import Dependencies

@Test func testDepInFunc() async throws {
    withDependencies {
        $0.date.now = .init(timeIntervalSince1970: 0)
    } operation: {
        #expect(Model().foo() == "1970-01-01 00:00:00 +0000")  // OK
    }
}

@Test func testDepInStaticFunc() async throws {
    withDependencies {
        $0.date.now = .init(timeIntervalSince1970: 0)
    } operation: {
        #expect(Model.foo() == "1970-01-01 00:00:00 +0000")  // OK
    }
}

@Test func testDepInFuncAssignment() async throws {
    let model = withDependencies {
        $0.date.now = .init(timeIntervalSince1970: 0)
    } operation: {
        Model()
    }

    #expect(model.foo() != "1970-01-01 00:00:00 +0000")  // <- 􀄵 Unimplemented: @Dependency(\.date)
}
```

So there must be something about our setup/invocation that messes up the `@Dependency` 🤔

The `􀄵 Unimplemented: @Dependency(\.date)` in the last test is actually great, because it would allow us to use the function-based `@Dependency` with confidence and we'd know when we're not setting it up correctly. Which perhaps is exactly what's happening right now 🤣

**Reply by finestructure** [reply]
It is (unsurprisingly) definitely an issue with our test setup, because when I pull the central bits of our test and the stuff under test into a standalone package, it works as expected. This is a Vapor project, so there are lots of test details I'm skipping when doing that. I'm guessing something we do as part of the Vapor setup is interfering. I'll try to pull more things into the standalone test to see if I can reproduce it outside out test machinery.

**Reply by mbrandonw** [reply]
We typically see this kind of problem when you accidentally link `Dependencies` to your test target. That creates duplicate symbols: https://swiftpackageindex.com/pointfreeco/swift-dependencies/main/documentation/dependencies/testing#Statically-linking-your-tests-target-to-Dependencies

Could that be happening in your situation?

**Reply by finestructure** [reply]
Mmm, this is a "pure" Swift package - there's no xcodeproj file where we'd be setting up special linking. It's all done via the `Package.swift` manifest. The `Dependencies` product is _only_ added to the app target's dependencies.

**Reply by finestructure** [reply]
FWIW, the error is the same if I run it via SwiftPM, i.e.

```
❯ swift test --filter AnalyzeErrorTests.test_analyze_refreshCheckout_failed
...
Build complete! (76.57s)
Test Suite 'Selected tests' started at 2024-10-03 18:45:22.627.
Test Suite 'SPI-ServerPackageTests.xctest' started at 2024-10-03 18:45:22.628.
Test Suite 'AnalyzeErrorTests' started at 2024-10-03 18:45:22.628.
Test Case '-[AppTests.AnalyzeErrorTests test_analyze_refreshCheckout_failed]' started.
/Users/sas/Projects/SPI/spi-server/.build/checkouts/swift-dependencies/Sources/Dependencies/DependencyValues/Date.swift:43: error: -[AppTests.AnalyzeErrorTests test_analyze_refreshCheckout_failed] : failed - Unimplemented: @Dependency(\.date)
Test Case '-[AppTests.AnalyzeErrorTests test_analyze_refreshCheckout_failed]' failed (1.598 seconds).
Test Suite 'AnalyzeErrorTests' failed at 2024-10-03 18:45:24.227.
	 Executed 1 test, with 1 failure (0 unexpected) in 1.598 (1.598) seconds
Test Suite 'SPI-ServerPackageTests.xctest' failed at 2024-10-03 18:45:24.227.
	 Executed 1 test, with 1 failure (0 unexpected) in 1.598 (1.598) seconds
Test Suite 'Selected tests' failed at 2024-10-03 18:45:24.227.
	 Executed 1 test, with 1 failure (0 unexpected) in 1.598 (1.600) seconds
```

**Reply by mbrandonw** [reply]
Are you able to get a minimal repro of some bad behavior so that we can take a look?

**Reply by finestructure** [reply]
I'm trying 😅 I'll have to extend my current test project with more of our setup. It doesn't reproduce there yet!

**Reply by finestructure** [reply]
Ok, I don't have a fully standalone reproducer yet, but I think I have a lead on what's going on. I've reduced the main function that is missing the `@Dependency` setup to

```swift
    static func analyze(database: Database, repository: Repository) async throws {
        try await database.transaction { tx in
            updateScore(repository: repository)
        }
    }
```

`updateScore` is using `@Dependency` and throws the `Unimplemented: @Dependency(\.date)` failure. You'll notice that `updateScore` is not actually using the Vapor db transaction at all - there's no saving happing.

If I remove the transaction wrapper like so

```swift
    static func analyze(database: Database, repository: Repository) async throws {
//        try await database.transaction { tx in
            updateScore(repository: repository)
//        }
    }
```

the error goes away. `transaction` is part of `fluent-kit`:

```swift
public extension Database {
    func transaction<T>(_ closure: @escaping @Sendable (any Database) async throws -> T) async throws -> T {
        try await self.transaction { db in
            self.eventLoop.makeFutureWithTask {
                try await closure(db)
            }
        }.get()
    }
}
```

and `makeFutureWithTask` is from `swift-nio`:

```swift
extension EventLoop {
    @inlinable
    public func makeFutureWithTask<Return>(
        _ body: @Sendable @escaping () async throws -> Return
    ) -> EventLoopFuture<Return> {
        let promise = self.makePromise(of: Return.self)
        promise.completeWithTask(body)
        return promise.futureResult
    }
}
```

with `completeWithTask`:

```swift
    @discardableResult
    @inlinable
    public func completeWithTask(_ body: @escaping @Sendable () async throws -> Value) -> Task<Void, Never> {
        Task {
            do {
                let value = try await body()
                self.succeed(value)
            } catch {
                self.fail(error)
            }
        }
    }
```

And that's the problem, isn't it? Isn't `@Dependencies` using some task local stuff and that's one of the things that unstructured `Task`s don't inherit?

**Reply by mbrandonw** [reply, marked as answer]
Hey @finestructure, ok I think I see the problem. It isn't the unstructured task, those specifically _do_ inherit task locals (only `Task.detached` does not).

However, _escaping_ closures lose dependencies. So by the time you get inside `transaction { … }` you have lost your dependencies, and similarly for `makeFutureWithTask { … }`. In general, escaping closures are the enemy of structured programming, and so therefore also the enemy of task locals or any of the tools in structured concurrency.

Luckily we do have a tool for this though. You can use [`withEscapedDependencies`](https://swiftpackageindex.com/pointfreeco/swift-dependencies/main/documentation/dependencies/withescapeddependencies(_:)-5xvi3) to escape dependencies in order to propagate them into an escaping closure:

```swift
static func analyze(database: Database, repository: Repository) async throws {
  withEscapedDependencies { dependencies in 
    try await database.transaction { tx in
      dependencies.yield {
        updateScore(repository: repository)
      }
    }
  }
}
```

I don't know if that's the exact place you need this tool, but that's the idea.

**Reply by finestructure** [reply]
Ah, that's a big relief that there might be a fix we can apply! I'll try this first thing in the morning and report back. Thank you!

**Reply by finestructure** [reply]
That's amazing - that fixed all our tests. Thanks so much, @mbrandonw !
